File: operators/cmd/crownlabs-image-list/update-crownlabs-image-list.py
#!/usr/bin/env python3

import argparse
import logging
import sched
import time

import grequests
import kubernetes
import requests


class ImageListUpdater:
    """
    This class periodically requests of the list of images from the Docker registy
    and saves the obtained information as a Kubernetes object.
    """

    def __init__(self, image_list_requestor, image_list_saver, registry_adv_name):
        """
        Initializes the object.
        :param image_list_requestor: the handler used to request the list of images from the Docker registry.
        :param image_list_saver: the handler used to save the retrieved information as a Kubernetes object.
        :param registry_adv_name: the Docker registry host name advertised in the ImageList object.
        """

        self.image_list_requestor = image_list_requestor
        self.image_list_saver = image_list_saver
        self.registry_adv_name = registry_adv_name
        self.scheduler = sched.scheduler(time.time, time.sleep)

    def run_update_process(self, interval):
        """
        Starts the scheduler loop to request and save the image list.
        :param interval: The interval (in seconds) between one execution and the following.
        """

        self._run_periodically(interval, self.update)
        self.scheduler.run()

    def update(self):
        """
        Performs the actual update process.
        """

        start = time.time()
        logger.debug("Starting the update process")

        try:
            # Obtain the list of images from the Docker registry
            images = self.image_list_requestor.get_image_list()
        except (requests.exceptions.RequestException, ValueError):
            logger.exception("Failed to retrieve data from upstream")
            return

        try:
            self.image_list_saver.update_image_list({
                "registryName": self.registry_adv_name,
                "images": ImageListUpdater.__process_image_list(images),
            })
        except kubernetes.client.rest.ApiException:
            logger.exception("Failed to save data as ImageList")
            return

        logger.info(f"Update process correctly completed in {time.time() - start:.2f} seconds")

    def _run_periodically(self, interval, action, *args, **kwargs):
        """
        Runs a given action periodically.
        :param interval: the interval between multiple executions (in seconds).
        :param action: the function to be executed periodically.
        :param kwargs: keyworded arguments passed to the action function.
        """

        logger.debug(f"Executing '{action.__name__}': next scheduled in {interval} seconds")

        # Schedule the next execution
        self.scheduler.enter(
            delay=interval, priority=1, action=self._run_periodically,
            argument=(interval, action), kwargs=kwargs)

        # Execute the action
        action(**kwargs)

    @staticmethod
    def __process_image_list(images):
        """
        Processes the list of images returned from upstream to remove the "latest" tags
        and converts it to the correct format expected by Kubernetes.
        :param images: the list of images retrieved from the Docker registry.
        :return: a list of images, suitable to be saved as ImageList.
        """

        converted_images = []

        for image in images:

            # Get the available tags
            versions = image.get("tags") or []

            # Remove the "latest" tag
            try:
                versions.remove("latest")
            except ValueError:
                pass

            # Are there still any tags?
            if versions:
                converted_images.append({
                    "name": image.get("name"),
                    "versions": versions,
                })

        return converted_images


class ImageListRequestor:
    """
    This class interacts with the docker registry to get the list of images currently available.
    """

    def __init__(self, url, username=None, password=None):
        """
        Initializes the object.
        :param url: the url used to contact the Docker registry.
        :param username: the username used to access the Docker registry (optional).
        :param password: the password used to access the Docker registry (optional).
        """

        self.url = url
        self.auth = (username, password)

    def get_image_list(self):
        """
        Performs the requests upstream to retrieve the information about the images.
        :return: an object representing the information retrieved.
        """

        logger.debug("Requesting the registry catalog upstream")
        repositories = self._do_single_get(
            ImageListRequestor.__get_catalog_path()
        ).get("repositories", [])

        logger.debug("Requesting the image details upstream")
        images = self._do_parallel_gets(
            ImageListRequestor.__map_repositories_to_paths(repositories)
        )

        return images

    def _do_single_get(self, path):
        """
        Performs a GET to the target path and returns the result.
        :param path: the path to be retrieved.
        :return: the json object extracted from the response.
        """

        return requests.get(url=f"{self.url}{path}", auth=self.auth).json()

    def _do_parallel_gets(self, paths):
        """
        Performs a set of parallel GETs to the target paths and returns the results.
        :param paths: the paths to be retrieved.
        :return: the set of json object extracted from the response.
        """

        requests = (grequests.get(f"{self.url}{path}", auth=self.auth) for path in paths)
        return (response.json() for response in grequests.imap(requests))

    @staticmethod
    def __get_catalog_path():
        """
        Returns the URL path corresponding to the catalog.
        :return: the URL path corresponding to the catalog.
        """
        return "/v2/_catalog"

    @staticmethod
    def __map_repositories_to_paths(repositories):
        """
        Returns the URL paths to obtain detailed information about the repositories.
        :param repositories: the set of repositories of interest.
        :return: the URL paths corresponding to the input repositories.
        """
        return (f"/v2/{repo}/tags/list" for repo in repositories)


class ImageListSaver:
    """
    Saves the list of images retrieved from the Docker registry as a Kubernetes object.
    """

    def __init__(self, name):
        """
        Initializes the object and loads the Kubernetes configuration.
        :param name: The name assigned to the ImageList resource.
        """

        self.name = name

        try:
            # Configuration loaded from a kube config
            kubernetes.config.load_kube_config()
        except kubernetes.config.config_exception.ConfigException:
            # Configuration loaded from within a pod
            kubernetes.config.load_incluster_config()

    def update_image_list(self, image_list_spec):
        """
        Updates the content or creates a new ImageList object.
        :param image_list_spec: the content of the ImageList object to be updated.
        """

        resource_version = self._get_image_list_version()
        if resource_version:
            self._update_image_list(image_list_spec, resource_version)
        else:
            self._create_image_list(image_list_spec)

    def _get_image_list_version(self):
        """
        Gets the current version of the ImageList.
        :returns: the current version of the image list or None
        """

        api_instance = kubernetes.client.CustomObjectsApi()
        try:
            data = api_instance.get_cluster_custom_object(
                **ImageListSaver.__get_imagelist_args(), name=self.name
            )
        except kubernetes.client.rest.ApiException:
            return None

        resource_version = data.get("metadata", {}).get("resourceVersion", None)
        logger.debug(f"Retrieved ImageList resource version: '{resource_version}'")
        return resource_version

    def _create_image_list(self, image_list_spec):
        """
        Creates a new ImageList object.
        :param image_list_spec: the content of the ImageList object to be created.
        """

        api_instance = kubernetes.client.CustomObjectsApi()
        api_instance.create_cluster_custom_object(
            **ImageListSaver.__get_imagelist_args(),
            body=self._create_image_list_object(image_list_spec)
        )
        logger.debug(f"ImageList '{self.name}' correctly created'")

    def _update_image_list(self, image_list_spec, resource_version):
        """
        Updates an existing ImageList object.
        :param image_list_spec: the content of the ImageList object to be updated.
        :param resource_version: the version of the resource to be updated.
        """

        api_instance = kubernetes.client.CustomObjectsApi()
        api_instance.replace_cluster_custom_object(
            **ImageListSaver.__get_imagelist_args(), name=self.name,
            body=self._create_image_list_object(image_list_spec, resource_version)
        )
        logger.debug(f"ImageList '{self.name}' correctly updated'")

    def _create_image_list_object(self, image_list_spec, resource_version=None):
        """
        Creates a new ImageList object, given the spec body.
        :param image_list_spec: the content of the ImageList object to be created.
        :param resource_version: the version of the resource to be updated.
        :returns: the ImageList json representation.
        """

        return {
            "apiVersion": "crownlabs.polito.it/v1alpha1",
            "kind": "ImageList",
            "metadata": {
                "name": self.name,
                "resourceVersion": resource_version,
            },
            "spec": image_list_spec,
        }

    @staticmethod
    def __get_imagelist_args():
        """
        Returns the parameters describing the ImageList API.
        """

        return {"group": "crownlabs.polito.it", "version": "v1alpha1", "plural": "imagelists"}


# Initialize the logger object
logger = logging.getLogger("update-crownlabs-image-list")


def configure_logger():
    """
    Configures the logger object with the required configuration
    """

    logger.setLevel(logging.DEBUG)

    formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")

    console_handler = logging.StreamHandler()
    console_handler.setLevel(logging.DEBUG)
    console_handler.setFormatter(formatter)

    logger.handlers.clear()
    logger.addHandler(console_handler)


if __name__ == "__main__":

    # Configure the logger
    configure_logger()

    # Parse the command line arguments
    parser = argparse.ArgumentParser(
        description="Periodically requests the list of images from a Docker registry and stores it as a Kubernetes CR")

    parser.add_argument("--advertised-registry-name", required=True,
                        help="the host name of the Docker registry where the images can be retrieved")
    parser.add_argument("--image-list-name", required=True,
                        help="the name assigned to the resulting ImageList object")
    parser.add_argument("--registry-url", required=True, help="the URL used to contact the Docker registry")
    parser.add_argument("--registry-username", help="the username used to access the Docker registry")
    parser.add_argument("--registry-password", help="the password used to access the Docker registry")
    parser.add_argument("--update-interval", required=True, type=int,
                        help="the interval (in seconds) between one update and the following")

    args = parser.parse_args()

    # Create the object reading the list of the images from the Docker registry
    logger.info(f"Upstream Docker registry: '{args.registry_url}' - Username: '{args.registry_username}'")
    image_list_requestor = ImageListRequestor(args.registry_url, args.registry_username, args.registry_password)

    # Create the object saving the retrieved information as a Kubernetes object
    logger.info(f"Target ImageList object: '{args.image_list_name}'")
    image_list_saver = ImageListSaver(args.image_list_name)

    # Create the object periodically perforing the update process
    image_list_updater = ImageListUpdater(image_list_requestor, image_list_saver, args.advertised_registry_name)

    logger.info("Starting the update process")
    try:
        image_list_updater.run_update_process(args.update_interval)
    except KeyboardInterrupt:
        logger.info("Received stop signal. Exiting")

File: operators/cmd/crownlabs-image-list/update-crownlabs-image-list.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"sync"

	"log"
	"os/signal"
	"syscall"
	"time"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
	"k8s.io/klog/v2"
)

// "k8s.io/client-go/kubernetes"
type ImageListRequestor struct {
	URL      string
	Username string
	Password string
	client   *http.Client
}

func NewImageListRequestor(url, username, password string) *ImageListRequestor {
	return &ImageListRequestor{
		URL:      url,
		Username: username,
		Password: password,
		client:   &http.Client{},
	}
}

func (r *ImageListRequestor) getImageList() ([]map[string]interface{}, error) {
	repositories, err := r.doSingleGet(r.getCatalogPath())
	if err != nil {
		return nil, err
	}
	repos, ok := repositories["repositories"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected catalog format")
	}

	paths := r.mapRepositoriesToPaths(repos)
	return r.doParallelGets(paths)
}

func (r *ImageListRequestor) doSingleGet(path string) (map[string]interface{}, error) {
	req, err := http.NewRequest("GET", r.URL+path, nil)
	if err != nil {
		return nil, err
	}
	if r.Username != "" && r.Password != "" {
		req.SetBasicAuth(r.Username, r.Password)
	}
	resp, err := r.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}
	return result, nil
}

func (r *ImageListRequestor) doParallelGets(paths []string) ([]map[string]interface{}, error) {
	var wg sync.WaitGroup
	results := make([]map[string]interface{}, len(paths))
	errors := make([]error, len(paths))

	for i, path := range paths {
		wg.Add(1)
		go func(i int, path string) {
			defer wg.Done()
			resp, err := r.doSingleGet(path)
			if err != nil {
				errors[i] = err
				return
			}
			results[i] = resp
		}(i, path)
	}
	wg.Wait()

	for _, err := range errors {
		if err != nil {
			return nil, err
		}
	}
	return results, nil
}

func (r *ImageListRequestor) getCatalogPath() string {
	return "/v2/_catalog"
}

func (r *ImageListRequestor) mapRepositoriesToPaths(repositories []interface{}) []string {
	paths := make([]string, len(repositories))
	for i, repo := range repositories {
		paths[i] = fmt.Sprintf("/v2/%v/tags/list", repo)
	}
	return paths
}

/////////////////////Image Saver start //////////////////////

type ImageListSaver struct {
	Name      string
	Client    dynamic.Interface
	GVR       schema.GroupVersionResource
	Namespace string // Not used for cluster-scoped resources, but kept for extensibility
}

func NewImageListSaver(name string) (*ImageListSaver, error) {
	var config *rest.Config
	var err error

	// Try kubeconfig first, then fallback to in-cluster config
	kubeconfig := os.Getenv("KUBECONFIG")
	if kubeconfig != "" {
		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
	} else {
		config, err = rest.InClusterConfig()
		if err != nil {
			// fallback to default kubeconfig location
			config, err = clientcmd.BuildConfigFromFlags("", clientcmd.RecommendedHomeFile)
		}
	}
	if err != nil {
		return nil, fmt.Errorf("failed to load kube config: %w", err)
	}

	client, err := dynamic.NewForConfig(config)
	if err != nil {
		return nil, fmt.Errorf("failed to create dynamic client: %w", err)
	}

	gvr := schema.GroupVersionResource{
		Group:    "crownlabs.polito.it",
		Version:  "v1alpha1",
		Resource: "imagelists",
	}

	return &ImageListSaver{
		Name:   name,
		Client: client,
		GVR:    gvr,
	}, nil
}

func (s *ImageListSaver) UpdateImageList(imageListSpec map[string]interface{}) error {
	resourceVersion, err := s.getImageListVersion()
	if err != nil {
		return err
	}
	if resourceVersion != "" {
		return s.updateImageList(imageListSpec, resourceVersion)
	}
	return s.createImageList(imageListSpec)
}

func (s *ImageListSaver) getImageListVersion() (string, error) {
	obj, err := s.Client.Resource(s.GVR).Get(context.TODO(), s.Name, metav1.GetOptions{})
	if err != nil {
		// Not found is not an error, just means it doesn't exist yet
		return "", nil
	}
	metadata, found, _ := unstructured.NestedMap(obj.Object, "metadata")
	if !found {
		return "", nil
	}
	if rv, ok := metadata["resourceVersion"].(string); ok {
		return rv, nil
	}
	return "", nil
}

func (s *ImageListSaver) createImageList(imageListSpec map[string]interface{}) error {
	obj := s.createImageListObject(imageListSpec, "")
	_, err := s.Client.Resource(s.GVR).Create(context.TODO(), obj, metav1.CreateOptions{})
	if err != nil {
		return fmt.Errorf("failed to create ImageList: %w", err)
	}
	fmt.Printf("ImageList '%s' created\n", s.Name)
	return nil
}

func (s *ImageListSaver) updateImageList(imageListSpec map[string]interface{}, resourceVersion string) error {
	obj := s.createImageListObject(imageListSpec, resourceVersion)
	_, err := s.Client.Resource(s.GVR).Update(context.TODO(), obj, metav1.UpdateOptions{})
	if err != nil {
		return fmt.Errorf("failed to update ImageList: %w", err)
	}
	fmt.Printf("ImageList '%s' updated\n", s.Name)
	return nil
}

func (s *ImageListSaver) createImageListObject(imageListSpec map[string]interface{}, resourceVersion string) *unstructured.Unstructured {
	obj := map[string]interface{}{
		"apiVersion": "crownlabs.polito.it/v1alpha1",
		"kind":       "ImageList",
		"metadata": map[string]interface{}{
			"name": s.Name,
		},
		"spec": imageListSpec,
	}
	if resourceVersion != "" {
		obj["metadata"].(map[string]interface{})["resourceVersion"] = resourceVersion
	}
	return &unstructured.Unstructured{Object: obj}
}

// ImageListUpdater periodically requests the list of images from the Docker registry
// and saves the obtained information as a Kubernetes object.
type ImageListUpdater struct {
	ImageListRequestor *ImageListRequestor
	ImageListSaver     *ImageListSaver
	RegistryAdvName    string
	ImageListPrefix    string // Prefix for the ImageList name, if needed
	// Scheduler removed; using goroutine and time.Ticker for scheduling
}

// NewImageListUpdater initializes the object.
func NewImageListUpdater(imageListRequestor *ImageListRequestor, imageListSaver *ImageListSaver, imageListPrefix string, registryAdvName string) *ImageListUpdater {
	return &ImageListUpdater{
		ImageListPrefix:    imageListPrefix,
		ImageListRequestor: imageListRequestor,
		ImageListSaver:     imageListSaver,
		RegistryAdvName:    registryAdvName,
	}
}

// RunUpdateProcess starts the scheduler loop to request and save the image list.
func (u *ImageListUpdater) RunUpdateProcess(interval time.Duration, stopCh <-chan struct{}) {
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			u.Update()
		case <-stopCh:
			return
		}
	}
}

// Update performs the actual update process.
func (u *ImageListUpdater) Update() {
	start := time.Now()
	log.Println("Starting the update process")

	images, err := u.ImageListRequestor.getImageList()
	if err != nil {
		log.Printf("Failed to retrieve data from upstream: %v", err)
		return
	}

	// spec := map[string]interface{}{
	// 	"registryName": u.RegistryAdvName,
	// 	"images":       processImageList(images),
	// }

	// if err := u.ImageListSaver.UpdateImageList(spec); err != nil {
	// 	log.Printf("Failed to save data as ImageList: %v", err)
	// 	return
	// }

	speccrownlabs_containerdisks := map[string]interface{}{
		"registryName": "crownlabs-containerdisks",
		"images":       processImageList(images),
	}
	u.ImageListSaver, err = NewImageListSaver("crownlabs-containerdisks")
	if err != nil {
		fmt.Printf("Error creating ImageListSaver: %v\n", err)
		os.Exit(1)
	}
	if err := u.ImageListSaver.UpdateImageList(speccrownlabs_containerdisks); err != nil {
		log.Printf("Failed to save data as ImageList: %v", err)
		return
	}

	spec_crownlabs_container_envs := map[string]interface{}{
		"registryName": "crownlabs-container-envs",
		"images":       processImageList(images),
	}

	u.ImageListSaver, err = NewImageListSaver("crownlabs-container-envs")
	if err != nil {
		fmt.Printf("Error creating ImageListSaver: %v\n", err)
		os.Exit(1)
	}
	if err := u.ImageListSaver.UpdateImageList(spec_crownlabs_container_envs); err != nil {
		log.Printf("Failed to save data as ImageList: %v", err)
		return
	}

	specCloadVm := map[string]interface{}{
		"registryName": "crownlabs-standalone",
		"images":       processImageList(images),
	}
	u.ImageListSaver, err = NewImageListSaver("crownlabs-standalone")
	if err != nil {
		fmt.Printf("Error creating ImageListSaver: %v\n", err)
		os.Exit(1)
	}

	if err := u.ImageListSaver.UpdateImageList(specCloadVm); err != nil {
		log.Printf("Failed to save data as ImageList: %v", err)
		return
	}
	log.Printf("Update process correctly completed in %.2f seconds", time.Since(start).Seconds())
}

// processImageList processes the list of images returned from upstream to remove the "latest" tags
// and converts it to the correct format expected by Kubernetes.
func processImageList(images []map[string]interface{}) []map[string]interface{} {
	var convertedImages []map[string]interface{}
	for _, image := range images {
		name, _ := image["name"].(string)
		tagsIface, ok := image["tags"]
		var versions []string
		if ok {
			switch tags := tagsIface.(type) {
			case []interface{}:
				for _, t := range tags {
					if tagStr, ok := t.(string); ok && tagStr != "latest" {
						versions = append(versions, tagStr)
					}
				}
			case []string:
				for _, tagStr := range tags {
					if tagStr != "latest" {
						versions = append(versions, tagStr)
					}
				}
			}
		}
		if len(versions) > 0 {
			convertedImages = append(convertedImages, map[string]interface{}{
				"name":     name,
				"versions": versions,
			})
		}
	}
	return convertedImages
}
func main() {
	var advertisedRegistryName, imageListName, registryURL string
	var updateInterval int

	flag.StringVar(&advertisedRegistryName, "advertised-registry-name", "", "The host name of the Docker registry where the images can be retrieved")
	flag.StringVar(&imageListName, "image-list-name", "", "The name assigned to the resulting ImageList object ")
	flag.StringVar(&registryURL, "registry-url", "", "The URL used to contact the Docker registry")
	flag.IntVar(&updateInterval, "update-interval", -1, "the interval (in seconds) between one update and the following")

	klog.InitFlags(nil)
	flag.Parse()

	if advertisedRegistryName == "" || imageListName == "" || registryURL == "" || updateInterval == -1 {
		fmt.Printf(`Usage: %s
		--advertised-registry-name <registry-name>
		--image-list-name <list-name>
		--registry-url <url>
		--update-interval <seconds>`, os.Args[0])
		fmt.Println()
		os.Exit(1)
	}

	// Example: create ImageListRequestor using parsed arguments
	imageListRequestor := NewImageListRequestor(registryURL, os.Getenv("REGISTRY_USERNAME"), os.Getenv("REGISTRY_PASSWORD"))

	// Example usage: get image list
	imageList, err := imageListRequestor.getImageList()
	if err != nil {
		fmt.Printf("Error retrieving image list: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Retrieved image list: %+v\n", imageList)

	fmt.Printf("Target ImageList object: '%s'\n", imageListName)
	imageListSaver, err := NewImageListSaver(imageListName)
	if err != nil {
		fmt.Printf("Error creating ImageListSaver: %v\n", err)
		os.Exit(1)
	}
	imageListUpdater := NewImageListUpdater(imageListRequestor, imageListSaver, imageListName, advertisedRegistryName)

	log.Println("Starting the update process")
	stopCh := make(chan struct{})
	go imageListUpdater.RunUpdateProcess(time.Duration(updateInterval)*time.Second, stopCh)

	// Wait for interrupt signal to gracefully shutdown
	c := make(chan os.Signal, 1)
	// import "os/signal" and "syscall" at the top
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	<-c
	log.Println("Received stop signal. Exiting")
	close(stopCh)
}
